core/templates/services/platform-feature.service.ts

```diff
@@ -0,0 +1,285 @@
+    try {
+      const item = this.loadSavedResults();
+      if (item && this.validateSavedResults(item)) {
+        PlatformFeatureService.featureStatusSummary = item.featureStatusSummary;
+        this.saveResults();
+      } else {
+        this.clearSavedResults();
+      }
+
+      if (!PlatformFeatureService.featureStatusSummary) {
+        PlatformFeatureService.featureStatusSummary = await this
+          .loadFeatureFlagsFromServer();
+        this.saveResults();
+      }
+    } catch (err) {
+      // If any error, just disable all features.
+      PlatformFeatureService.featureStatusSummary = this
+        .featureStatusSummaryObjectFactory.createDefault();
+      PlatformFeatureService._initializedWithError = true;
+      this.clearSavedResults();
+    }
+  }
+
+  private async loadFeatureFlagsFromServer(): Promise<FeatureStatusSummary> {
+    const context = this.generateClientContext();
+    return this.platformFeatureBackendApiService.fetchFeatureFlags(context);
+  }
+
+  /**
+   * Saves the results in sessionStorage, along with current timestamp and
+   * the current session id.
+   */
+  private saveResults(): void {
+    const item = {
+      timestamp: this.getCurrentTimestamp(),
+      sessionId: this.getSessionIdFromCookie(),
+      featureStatusSummary: PlatformFeatureService.featureStatusSummary
+        .toBackendDict(),
+    };
+    this.windowRef.nativeWindow.sessionStorage.setItem(
+      PlatformFeatureService.SESSION_STORAGE_KEY, JSON.stringify(item));
+  }
+
+  /**
+   * Clears results from the sessionStorage, if any.
+   */
+  private clearSavedResults(): void {
+    this.windowRef.nativeWindow.sessionStorage.removeItem(
+      PlatformFeatureService.SESSION_STORAGE_KEY);
+  }
+
+  /**
+   * Reads and parse feature flag results from the sessionStorage.
+   *
+   * @returns {FeatureFlagsCacheItem|null} - Saved results along with timestamp
+   * and session id. Null if there isn't any saved result.
+   */
+  private loadSavedResults(): FeatureFlagsCacheItem | null {
+    const savedStr = this.windowRef.nativeWindow.sessionStorage.getItem(
+      PlatformFeatureService.SESSION_STORAGE_KEY);
+    if (savedStr) {
+      const savedObj = JSON.parse(savedStr);
+      return {
+        timestamp: savedObj.timestamp,
+        sessionId: savedObj.sessionId,
+        featureStatusSummary: (
+          this.featureStatusSummaryObjectFactory.createFromBackendDict(
+            savedObj.featureStatusSummary))
+      };
+    }
+    return null;
+  }
+
+  /**
+   * Validates the result saved in sessionStorage.
+   *
+   * @param {FeatureFlagsCacheItem} item - The result item loaded from
+   * sessionStorage.
+   *
+   * @returns {boolean} - True if the result is valid and can be directly used.
+   */
+  private validateSavedResults(item: FeatureFlagsCacheItem): boolean {
+    if (this.getCurrentTimestamp() - item.timestamp >
+        PlatformFeatureService.SESSION_STORAGE_CACHE_TTL) {
```

Comment:
Suggest adding tests specifically to make sure only the outdated parameters are updated since this is an important behavior to ensure.

Reply:
Per the code and your design doc, my understanding was that we track individual parameters for being outdated. I recall a comment in your design doc indicating that only outdated feature flags are updated when they reach their TTL threshold. My suggestion is to ensure there's a test to verify this (e.g. if you add one feature at time X, advance the clock, add another feature Y, and advance the clock such that X is expired but Y is not, X should be updated).